rfd/xxxx-azure-node-join.md

```diff
@@ -0,0 +1,130 @@
+
+This is the Azure equivalent of EC2/IAM join as described in
+[RFD 41](https://github.com/gravitational/teleport/blob/master/rfd/0041-aws-node-join.md).
+
+## Why
+
+This will make provisioning nodes on Azure simpler and arguably more secure.
+There is no need to configure static tokens, or to manage the provisioning and
+deployment of dynamic tokens.
+
+## Details
+
+In place of a join token, nodes will present a signed attested data document
+to the auth server. The document will be fetched on the nodes via instance
+metadata. The request can include a user-specified nonce which will also be
+signed. Like the custom headers in IAM join, the nonce allows Teleport to issue
+a challenge (crypto random string of bytes) that the node must include in the
+signed request. We can use gRPC bidirectional streams to implement the
+following flow at a new `RegisterWithAzureToken` gRPC endpoint:
+
+1. The Node initiates a `RegisterWithAzureToken` request.
+2. The auth server enforces that TLS is used for transport.
+3. The auth server sends a base64 encoded 24 byte crypto-random challenge.
+  - We use 24 bytes instead of 32 because the nonce parameter is limited to 32
+    characters, and 24 bytes require 32 base64 characters.
+4. The node requests the attested data document using the challenge as the
+  nonce. Azure returns the signed document.
+5. The node sends a join request to the auth server, including the signed
+  document and other paramaters in a `RegisterUsingTokenRequest`.
+  - `RegisterUsingTokenRequest` will need to be extended to include info not in
+  the attested data document (notably the VM's resource group and region).
+6. The auth server checks that signed document is valid:
+  - The document's pkcs7 signature is valid.
+  - The returned nonce matches the issued challenge.
+  - The requested token allows the node to join.
+7. The auth server sends credentials to join the cluster.
+
+As with IAM join, the flow will occur within a single streaming gRPC request and
+only 1 attempt with a given challenge will be allowed, with a 1 minute timeout.
+
+### Teleport Configuration
+
+The existing provision token type can be extended to support Azure
+authentication, using new Azure-specific fields in the token rules section.
+
+```yaml
+kind: token
+version: v2
+metadata:
+  name: example_azure_token
+spec:
+  roles: [Node,Kube,Db]
+  allow:
+    # new token fields below
+
+    # Subscription from which nodes can join. Required.
+    - azure_subscription: "22222222"
+
+      # Resource groups from which nodes can join. If empty or omitted, nodes
+      # from any resource group are allowed.
+      azure_resource_groups: ["rg1", "rg2"]
+      
+      # Regions from which nodes can join. If empty or omitted, nodes from any
+      # region are allowed.
+      azure_regions: ["r1", "r2"]
+
+  # The duration after the attested data created time that the document will
+  # be accepted. The default is 5 minutes.
+  azure_attested_data_ttl: 5m
```

Comment:
I don't think this needs to be included. The `aws_iid_ttl` is included for the ec2 join method to limit the usefulness of an exfiltrated IID. Since this design includes a challenge-response, and the entire exchange has a 1 minute timeout, we already know that the attested data is new (because it includes the challenge) and an exfiltrated attested data document would be of no use without the correct challenge.